ttyplot.c

```diff
@@ -289,9 +290,99 @@ void redraw_screen(const char * errstr) {
     refresh();
 }
 
+// Handle a single value from the input stream.
+// Return whether we got a full data record.
+int handle_value(double value)
+{
+    static double saved_value;
+    static int saved_value_valid = 0;
+
+    // First value of a 2-value record: save it for later.
+    if (two && !saved_value_valid) {
+        saved_value = value;
+        saved_value_valid = 1;
+        return 0;
+    }
+
+    // Otherwise we have a full record.
+    n = (n+1) % plotwidth;
+    if (two) {
+        values1[n] = saved_value;
+        values2[n] = value;
+        saved_value_valid = 0;
+    } else {
+        values1[n] = value;
+    }
+    if (rate)
+        td = derivative(&values1[n], two ? &values2[n] : NULL, &now);
+    return 1;
+}
+
+// Parse the input data. Return the number of bytes consumed.
+size_t parse_input(char *buffer, size_t length)
+{
+    static const char delimiters[] = " \t\r\n";  // white space
+
+    // Find the last delimiter.
+    int end;
+    for (end = length - 1; end >= 0; end--) {
+        if (strchr(delimiters, buffer[end]))  // found
+            break;
+    }
+    if (end < 0)  // no delimiter
+        return 0;
+    buffer[end] = '\0';
+
+    // Tokenize and parse.
+    int records = 0;  // number or records found
+    char *str = buffer;
+    char *token;
+    while ((token = strtok(str, delimiters)) != NULL) {
+        records += handle_value(strtod(token, NULL));
+        str = NULL;  // tell strtok() to stay on the same string
+    }
+    v += records;
+    if (records > 0)
+        redraw_screen(errstr);
+    return end + 1;
+}
+
+// Handle an "input ready" event, where only a single read() is guaranteed to not block.
+// Return whether the input stream got closed.
+int handle_input(void)
```

Comment:
- This `int` is another `bool` in disguise. Let's use `bool` please.
- (Function `handle_input` has two jobs rather than one: (1) push stdin into the buffer, and (2) process the buffer content wrapping the call to `parse_input`. Maybe it cleaner splitting it up in two.)

Reply:
Fixed the `int` → `bool` change.

> Function `handle_input` has two jobs rather than one:

It has to fully handle the event.

> (1) push stdin into the buffer, and (2) process the buffer content wrapping the call to `parse_input`.

(3) shift the consumed data out of the buffer.

> Maybe it cleaner splitting it up in two.

That's what I did by factoring `parse_input` out of this function. I do not see how to split it further without that feeling unnatural.